aws/s3/s3.go

```diff
@@ -0,0 +1,369 @@
+func (s *S3) Put(bucket, key string, object []byte) error {
+	input := s3.PutObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+		Body:   bytes.NewReader(object),
+	}
+
+	_, err := s.client.PutObject(context.TODO(), &input)
+	return err
+}
+
+// Put puts the object in bucket with metadata using specified key.
+func (s *S3) PutWithMetadata(bucket, key string, object []byte, metadata Meta) error {
+	input := s3.PutObjectInput{
+		Bucket:   aws.String(bucket),
+		Key:      aws.String(key),
+		Body:     bytes.NewReader(object),
+		Metadata: metadata,
+	}
+
+	_, err := s.client.PutObject(context.TODO(), &input)
+	return err
+}
+
+// Exists checks if an object for key already exists in the bucket.
+func (s *S3) Exists(bucket, key string) (bool, error) {
+	input := s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+
+	_, err := s.client.HeadObject(context.TODO(), &input)
+	if err == nil {
+		return true, nil
+	}
+
+	var nf *types.NotFound
+	if errors.As(err, &nf) {
+		return false, nil
+	}
+	return false, err
+}
+
+// List returns a list of object keys that match the provided prefix.
+// It will not return more than the specified max number of keys.
+// Keys are in alphabetical order.
+func (s *S3) List(bucket, prefix string, max int32) ([]string, error) {
+	input := s3.ListObjectsV2Input{
+		Bucket:  aws.String(bucket),
+		Prefix:  aws.String(prefix),
+		MaxKeys: max,
+	}
+
+	out, err := s.client.ListObjectsV2(context.TODO(), &input)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]string, 0)
+	for _, o := range out.Contents {
+		result = append(result, *o.Key)
+	}
+	return result, nil
+}
+
+// ListAll returns a list of ALL object keys that match the provided prefix.
+// Keys are in alphabetical order.
+func (s *S3) ListAll(bucket, prefix string) ([]string, error) {
+
+	result := make([]string, 0)
+
+	allFound := false
+	var continuationToken *string
+
+	for !allFound {
+
+		input := s3.ListObjectsV2Input{
+			Bucket:            aws.String(bucket),
+			Prefix:            aws.String(prefix),
+			ContinuationToken: continuationToken,
+		}
+
+		out, err := s.client.ListObjectsV2(context.TODO(), &input)
+		if err != nil {
+			return nil, err
+		}
+		for _, o := range out.Contents {
+			result = append(result, *o.Key)
+		}
+		// When result is not truncated, it means all matching keys have been found.
+		if !out.IsTruncated {
+			allFound = true
+		} else {
+			continuationToken = out.NextContinuationToken
+		}
+	}
+	return result, nil
+}
+
+// Returns whether there is an object in bucket with specified prefix.
+func (s *S3) PrefixExists(bucket, prefix string) (bool, error) {
+	input := s3.ListObjectsV2Input{
+		Bucket:  aws.String(bucket),
+		Prefix:  aws.String(prefix),
+		MaxKeys: 1,
+	}
+	out, err := s.client.ListObjectsV2(context.TODO(), &input)
+	if err != nil {
+		return false, err
+	}
+	if len(out.Contents) > 0 {
+		return true, nil
+	} else {
```

Comment:
This `else` is redundant, i.e. if the first "if" holds the function has returned already and the flow would have stopped.